pkg/ipmi/ipmi.go

```diff
@@ -173,3 +191,94 @@ func (i *IPMI) ShutoffWatchdog() error {
 
 	return nil
 }
+
+func (i *IPMI) waitForSetInProgressCleared() error {
+	req := &req{}
+	req.msg.cmd = _GET_SYSTEM_INFO_PARAMETERS
+	req.msg.netfn = _IPMI_NETFN_APP
+	var data [4]byte
+
+	data[0] = 0
+	data[1] = _SET_IN_PROGRESS
+	data[2] = 0
+	data[3] = 0
+	req.msg.data = unsafe.Pointer(&data)
+	req.msg.dataLen = 4
+
+	retry := 20
+	for retry > 0 {
+		recv, err := i.sendrecv(req)
+		if err != nil {
+			return err
+		}
+		// response data byte 3 bit[1:0] == 00b indicates set complete
+		if len(recv) == 3 && (recv[2]&0x3) == 0 {
+			return nil
+		}
+		time.Sleep(1 * time.Millisecond)
+		retry--
+	}
+
+	return fmt.Errorf("Wait for Set in Progress cleared timeout")
+
+}
+
+func (i *IPMI) setsysinfo(data *setSystemInfoReq, dataLen uint16) error {
+	req := &req{}
+	req.msg.cmd = _SET_SYSTEM_INFO_PARAMETERS
+	req.msg.netfn = _IPMI_NETFN_APP
+	req.msg.dataLen = dataLen
+	req.msg.data = unsafe.Pointer(data)
+
+	if err := i.waitForSetInProgressCleared(); err != nil {
+		return err
+	}
+
+	if _, err := i.sendrecv(req); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SetSystemFWVersion sets the provided system firmware version to BMC via IPMI.
+func (i *IPMI) SetSystemFWVersion(version string) error {
+	strlenMax := 64 // Set 64 Bytes as the maximal version string length
```

Comment:
From where did you get this max? I think `ipmitool` allows 62 bytes, 4 writes of 14 the first and 16 the following ones.

Reply:
I referenced from the IPMI spec, for system firmware version it says for non-EFI system firmware it is recommended that four blocks (64 bytes) of storage be provided. For EFI-based systems 256 bytes is recommended.

Reply:
Yes but from those 4 blocks, the first block has 2 bytes being used for something else. Isn't that the case?

Reply:
The string length will not be dependent on the first block having 2 bytes being used for something else, because the for-loop will keep sending out the string data until the len is <= 0, the for-loop and the spec block format are both independent of the strlenMax value. 

I check the ipmitool code and it sets the maximal length to a fixed value when the string length exceeds the maximal length and continue sending out to BMC, I think it's better so I change my code and update the commit:
-       if len > strlenMax || len == 0 {
-               return fmt.Errorf("Version length is 0 or longer than the suggested maximal length %d", strlenMax)
+       if len == 0 {
+               return fmt.Errorf("Version length is 0")
+       } else if len > strlenMax {
+               len = strlenMax
        }

Reply:
> I check the ipmitool code and it sets the maximal length to a fixed value when the string length exceeds the maximal length

Which is that max value? Can you point me to it?

My understanding is that it only allows 4 writes/sets: https://github.com/ipmitool/ipmitool/blob/master/lib/ipmi_mc.c#L1412

And with 4 writes/sets the max value will be 32 bytes. Am I missing something?

Reply:
It's at
https://github.com/ipmitool/ipmitool/blob/master/lib/ipmi_mc.c#L1559
and you're right the ipmitool limits the maxmimal length to 62 bytes, but I limit it to 64 bytes according to the IPMI spec. Why does ipmitool limit to 62 but not 64 I don't know, maybe the author wants to limit the number of sets to 4, if 64 bytes it would take 5 sets to send, I can also limit it to 62 bytes I guess the 2 bytes don't matter that much.

Reply:
The spec mentions 4 blocks of storage and 64bytes which is misleading to me. What does it mean with 4 blocks? Does it mean 4 writes? If so the limit would be 32 bytes. I would follow `ipmitool` :P

Reply:
Maybe in those 64 bytes it is also considering the encoding and string length, which will leave only 64 bytes for the string.

Reply:
Yes that's confusing, so I updated the commit to 62 bytes (4 sets/data blocks).